### PR TITLE
[Merged by Bors] - TY-2136 impl kpe models manually [1]

### PR DESCRIPTION
**Pull request by janpetschexain**
**References**

- [TY-2136]

**Summary**

i'd suggest to look at the [commit of the raw impl](https://github.com/xaynetwork/xayn_ai/blob/b692b68714776f3ddcaee70cd48f8e64875c28b8/layer/src/conv.rs) to check correctness wrt the original C++ version and then at the last commit to check if this was rustified correctly:
- add a 1D convolutional layer and implement the paths needed from the [C++ torch source](https://pytorch.org/docs/stable/generated/torch.nn.functional.conv1d.html) for the chosen configuration in the [KPE prototype](https://github.com/xaynetwork/xayn_ai_research/blob/main/experiments/preference_modeling/keyphrase_extraction/bert2joint/model.py#L193-L195)
- add tests, the original only had one test and i created more from running the original

**Todo**

- ~refactor the looped assignments as iterator collections whereever possible~

[TY-2136]: https://xainag.atlassian.net/browse/TY-2136?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: layer/src/conv.rs
Comment (rustonaut) [inline layer/src/conv.rs:11]:
```suggestion
    weights: Array3<f32>,
    bias: Option<Array1<f32>>,
```

I still need to double check it but bias being 3-D and with that higher dimensional then weights seems strange, tbh.

**Review by rustonaut — CHANGES_REQUESTED** [review]
Besides the one dimension mess up one problem is that the doc-links are not perma links, i.e.  they link to master and when master changes might point to pretty arbitrary places

**Comment by janpetschexain** [comment]
> Besides the one dimension mess up one problem is that the doc-links are not perma links, i.e. they link to master and when master changes might point to pretty arbitrary places

i want to remove/replace the doclinks once the review is done, they are just here to help reviewing the code.

File: layer/src/conv.rs
Comment (janpetschexain) [reply layer/src/conv.rs:9]: the weights are already reshaped into the dimension needed during the run, such that we don't have to reshape them every time.

the bias is just extended with singleton dimensions from `(bias_size,)` to `(1, bias_size, 1)` for the same reason.

Comment (rustonaut) [reply layer/src/conv.rs:9]: Oh, you are right I even preciously checked that the transformation was correct :see_no_evil:

File: layer/src/conv.rs
Comment (rustonaut) [inline layer/src/conv.rs]: is it ok to remove this?

**Review by rustonaut — COMMENTED** [review]
-

File: layer/src/conv.rs
Comment (janpetschexain) [reply layer/src/conv.rs]: yes, it's already implied: both `groups` and `channel_out_size` are guaranteed to be non-zero, hence any `channel_out_size < groups` will cause `channel_out_size % groups == channel_out_size != 0` wich is already checked below.

**Review by rustonaut — APPROVED** [review]

**Comment by janpetschexain** [comment]
@rustonaut why are you merging my pr?

**Comment by rustonaut** [comment]
Isn't it the new PR flow? Given that bors by default doesn't allow the PR opener to merge it?

**Comment by rustonaut** [comment]
I mane `borse r+` ==  “reviewed: looks good.”

**Comment by janpetschexain** [comment]
> Isn't it the new PR flow? Given that bors by default doesn't allow the PR opener to merge it?

why should it be? the pr opener can use bors to merge the pr as soon as the pr has one approval, see the list of merged prs of the repo. anything else would be pretty weird.

> I mane borse r+ == “reviewed: looks good.”

actually `bors r+` is just shorthand for `bors merge`, so it's not a "reviewed: looks good" but a "merge pr".

**Comment by rustonaut** [comment]
> actually bors r+ is just shorthand for bors merge, so it's not a "reviewed: looks good" but a "merge pr".

It is a quote from our own documentation on Confluence.

**Comment by janpetschexain** [comment]
> > actually bors r+ is just shorthand for bors merge, so it's not a "reviewed: looks good" but a "merge pr".
> 
> It is a quote from our own documentation on Confluence.

then we should fix that documentation.

**Pull request closed without merging**
